lib/auth/webauthn/messages_test.go

```diff
@@ -56,3 +57,270 @@ func TestCredentialAssertionResponse_json(t *testing.T) {
+			createCC: func() *wanlib.CredentialCreation {
+				cp := *okCC
+				cp.Response.User.DisplayName = ""
+				return &cp
+			},
+			wantErr: "user display name",
+		},
+		{
+			name: "user ID",
+			createCC: func() *wanlib.CredentialCreation {
+				cp := *okCC
+				cp.Response.User.ID = nil
+				return &cp
+			},
+			wantErr: "user ID",
+		},
+		{
+			name: "cc without rp name",
+			createCC: func() *wanlib.CredentialCreation {
+				cp := *okCC
+				cp.Response.RelyingParty.Name = ""
+				return &cp
+			},
+			wantErr: "relying party name required",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			err := test.createCC().Validate()
+			if test.wantErr != "" {
+				require.ErrorContains(t, err, test.wantErr, "Validate returned err = %q, want %q", err, test.wantErr)
+			} else {
+				require.NoError(t, err, "Validate returned err %v, want nil", err)
+			}
+		})
+	}
+}
+
+func TestCredentialAssertion_Validate(t *testing.T) {
+	okAssertion := &wanlib.CredentialAssertion{
+		Response: protocol.PublicKeyCredentialRequestOptions{
+			Challenge:      make([]byte, 32),
+			RelyingPartyID: "example.com",
+			AllowedCredentials: []protocol.CredentialDescriptor{
+				{Type: protocol.PublicKeyCredentialType, CredentialID: []byte{1, 2, 3, 4, 5}},
+			},
+		},
+	}
+
+	nilChallengeAssertion := *okAssertion
+	nilChallengeAssertion.Response.Challenge = nil
+
+	emptyRPIDAssertion := *okAssertion
+	emptyRPIDAssertion.Response.RelyingPartyID = ""
+	tests := []struct {
+		name      string
+		assertion *wanlib.CredentialAssertion
+		wantErr   string
+	}{
+		{
+			name:      "ok", // check that good params are good
+			assertion: okAssertion,
+			wantErr:   "",
+		},
+		{
+			name:    "nil assertion",
+			wantErr: "assertion required",
+		},
+		{
+			name:      "assertion without challenge",
+			assertion: &nilChallengeAssertion,
+			wantErr:   "challenge",
+		},
+		{
+			name:      "assertion without RPID",
+			assertion: &emptyRPIDAssertion,
+			wantErr:   "relying party ID",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			err := test.assertion.Validate()
+			if test.wantErr != "" {
+				require.ErrorContains(t, err, test.wantErr, "Validate returned err = %q, want %q", err, test.wantErr)
+			} else {
+				require.NoError(t, err, "Validate returned err %v, want nil", err)
+			}
+		})
+	}
+}
+
+func TestRequireResidentKey(t *testing.T) {
+	tests := []struct {
+		name    string
+		in      protocol.AuthenticatorSelection
+		want    bool
+		wantErr string
+	}{
+		{
+			name: "nothing set",
+			in:   protocol.AuthenticatorSelection{},
+			want: false,
+		},
+		{
+			name: "mismatch of ResidentKey and RequireResidentKey - return error",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        protocol.ResidentKeyRequirementDiscouraged,
+				RequireResidentKey: protocol.ResidentKeyRequired(),
+			},
+			wantErr: "invalid combinantion of ResidentKey: discouraged and RequireResidentKey: true",
```

Comment:
Suggestion: a shorter check here would do, makes the test a little bit less fragile.

```suggestion
			wantErr: "invalid combination of ResidentKey",
```